optuna/integration/tensorboard.py

```diff
@@ -0,0 +1,155 @@
+                        distribution_type, distribution_list
+                    )
+                )
+
+        for param_name, param_distribution in param_distributions_optuna_objects.items():
+
+            if isinstance(param_distribution, optuna.distributions.UniformDistribution):
+                self._hp_params[param_name] = hp.HParam(
+                    param_name, hp.RealInterval(param_distribution.low, param_distribution.high)
+                )
+            elif isinstance(param_distribution, optuna.distributions.LogUniformDistribution):
+                self._hp_params[param_name] = hp.HParam(
+                    param_name, hp.RealInterval(param_distribution.low, param_distribution.high)
+                )
+            elif isinstance(param_distribution, optuna.distributions.DiscreteUniformDistribution):
+                self._hp_params[param_name] = hp.HParam(
+                    param_name, hp.Discrete(param_distribution.low, param_distribution.high)
+                )
+            elif isinstance(param_distribution, optuna.distributions.IntUniformDistribution):
+                self._hp_params[param_name] = hp.HParam(
+                    param_name, hp.IntInterval(param_distribution.low, param_distribution.high)
+                )
+            elif isinstance(param_distribution, optuna.distributions.CategoricalDistribution):
+                self._hp_params[param_name] = hp.HParam(
+                    param_name, hp.Discrete(param_distribution.choices)
+                )
+            else:
+                distribution_list = [
+                    optuna.distributions.UniformDistribution.__name__,
+                    optuna.distributions.LogUniformDistribution.__name__,
+                    optuna.distributions.DiscreteUniformDistribution.__name__,
+                    optuna.distributions.IntUniformDistribution.__name__,
+                    optuna.distributions.CategoricalDistribution.__name__,
+                ]
+                raise NotImplementedError(
+                    "The distribution {} is not implemented. "
+                    "The parameter distribution should be one of the {}".format(
+                        param_distribution, distribution_list
+                    )
+                )
+
+    def __call__(self, study: "optuna.study.Study", trial: "optuna.trial.FrozenTrial") -> None:
+
+        trial_value = trial.value if trial.value is not None else float("nan")
+
+        hparams = dict()
+        for param_name, param_value in trial.params.items():
+            hparams[self._hp_params[param_name]] = param_value
+
+        run_name = "trial-%d" % trial.number
+        run_dir = os.path.join(self._dirname, run_name)
+
+        with tf.summary.create_file_writer(run_dir).as_default():
+            hp.hparams(hparams)  # record the values used in this trial
+            tf.summary.scalar(self._metric_name, trial_value, step=1)
```

Comment:
We can remove the `param_distributions` argument by introducing "lazy" initialization as follows.

```suggestion
    def __init__(self, dirname: str, metric_name: str) -> None:

        _imports.check()

        self._dirname = dirname
        self._metric_name = metric_name
        self._hp_params = dict()  # type: Dict[str, hp.HParam]

    def __call__(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial) -> None:

        if len(self._hp_params) == 0:
            self._initialization(study)

        trial_value = trial.value if trial.value is not None else float("nan")

        hparams = dict()
        for param_name, param_value in trial.params.items():
            if param_name not in self._hp_params:
                self._add_distributions(trial.distributions)
            hparams[self._hp_params[param_name]] = param_value

        run_name = "trial-%d" % trial.number
        run_dir = os.path.join(self._dirname, run_name)

        with tf.summary.create_file_writer(run_dir).as_default():
            hp.hparams(hparams)  # record the values used in this trial
            tf.summary.scalar(self._metric_name, trial_value, step=1)

    def _add_distributions(
        self, distributions: Dict[str, optuna.distributions.BaseDistribution]
    ) -> None:

        for param_name, param_distribution in distributions.items():
            if isinstance(param_distribution, optuna.distributions.UniformDistribution):
                self._hp_params[param_name] = hp.HParam(
                    param_name, hp.RealInterval(param_distribution.low, param_distribution.high)
                )
            elif isinstance(param_distribution, optuna.distributions.LogUniformDistribution):
                self._hp_params[param_name] = hp.HParam(
                    param_name, hp.RealInterval(param_distribution.low, param_distribution.high)
                )
            elif isinstance(param_distribution, optuna.distributions.DiscreteUniformDistribution):
                self._hp_params[param_name] = hp.HParam(
                    param_name, hp.Discrete(param_distribution.low, param_distribution.high)
                )
            elif isinstance(param_distribution, optuna.distributions.IntUniformDistribution):
                self._hp_params[param_name] = hp.HParam(
                    param_name, hp.IntInterval(param_distribution.low, param_distribution.high)
                )
            elif isinstance(param_distribution, optuna.distributions.CategoricalDistribution):
                self._hp_params[param_name] = hp.HParam(
                    param_name, hp.Discrete(param_distribution.choices)
                )
            else:
                distribution_list = [
                    optuna.distributions.UniformDistribution.__name__,
                    optuna.distributions.LogUniformDistribution.__name__,
                    optuna.distributions.DiscreteUniformDistribution.__name__,
                    optuna.distributions.IntUniformDistribution.__name__,
                    optuna.distributions.CategoricalDistribution.__name__,
                ]
                raise NotImplementedError(
                    "The distribution {} is not implemented. "
                    "The parameter distribution should be one of the {}".format(
                        param_distribution, distribution_list
                    )
                )

    def _initialization(self, study: optuna.Study) -> None:

        completed_trials = [
            trial
            for trial in study.get_trials(deepcopy=False)
            if trial.state == optuna.trial.TrialState.COMPLETE
        ]
        for trial in completed_trials:
            self._add_distributions(trial.distributions)
```